src/transformers/models/olmoe/modeling_olmoe.py

```diff
@@ -0,0 +1,1368 @@
+            key_states.clamp_(min=-self.config.clip_qkv, max=self.config.clip_qkv)
+            value_states.clamp_(min=-self.config.clip_qkv, max=self.config.clip_qkv)
+
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        cos, sin = self.rotary_emb(value_states, position_ids)
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
+
+        if past_key_value is not None:
+            # sin and cos are specific to RoPE models; cache_position needed for the static cache
+            cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
+
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+        causal_mask = attention_mask
+        # if attention_mask is not None and cache_position is not None:
+        if attention_mask is not None:
+            causal_mask = causal_mask[:, :, :, : key_states.shape[-2]]
+
+        # SDPA with memory-efficient backend is currently (torch==2.1.2) bugged with non-contiguous inputs with custom attn_mask,
+        # Reference: https://github.com/pytorch/pytorch/issues/112577.
+        if query_states.device.type == "cuda" and causal_mask is not None:
+            query_states = query_states.contiguous()
+            key_states = key_states.contiguous()
+            value_states = value_states.contiguous()
+
+        # We dispatch to SDPA's Flash Attention or Efficient kernels via this `is_causal` if statement instead of an inline conditional assignment
+        # in SDPA to support both torch.compile's dynamic shapes and full graph options. An inline conditional prevents dynamic shapes from compiling.
+        is_causal = True if causal_mask is None and q_len > 1 else False
+
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            is_causal=is_causal,
+        )
+
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.view(bsz, q_len, self.hidden_size)
+
+        attn_output = self.o_proj(attn_output)
+
+        return attn_output, None, past_key_value
+
+
+OLMOE_ATTENTION_CLASSES = {
+    "eager": OlmoeAttention,
+    "flash_attention_2": OlmoeFlashAttention2,
+    "sdpa": OlmoeSdpaAttention,
+}
+
+
+class OlmoeSparseMoeBlock(nn.Module):
```

Comment:
reviewing fast but probably missing copied from here as well no?

Reply:
(Qwen2-moe)

Reply:
also I probably would not recommend you to use this formulation. As it will not be compatible with torch compile. #30793 gives explanation but if you have a conversion script, could be good to default with the correct format (I mean format that will allow ud to use the torch compile compatible formulation!)

Reply:
Different from Qwen2-moe cuz we have no shared expert

Reply:
> also I probably would not recommend you to use this formulation. As it will not be compatible with torch compile. https://github.com/huggingface/transformers/pull/30793 gives explanation but if you have a conversion script, could be good to default with the correct format (I mean format that will allow ud to use the torch compile compatible formulation!)

There's no new & better formulation used by any model in transformers yet, right?

Reply:
It is implemented in the linked PR here: https://github.com/huggingface/transformers/pull/30793/files#diff-733ab0a772c69f78b1d8ed361e6ae1fda7243652887aed0bab5d3ecf07794c01R789

Reply:
it's completely up to you!

Reply:
Oh thanks; I guess I was confused why the PR got closed - has it not been merged in anywhere?

Reply:
I took a look but it seems like there's a trade-off where the PR from @zhenglongjiepheonix increases memory usage for better speed while this implementation here requires less memory but is slower? Can you correct me if I'm wrong @zhenglongjiepheonix @ArthurZucker ?

Reply:
It's mostly for compatibility of torch compile, the compiled model is 3x~4x faster in the case of Mixtral, have not made any comparisons on the eager mode

Reply:
That's great! But it requires more memory or not really?

Reply:
Yes, it does

Reply:
at least for training, I think the extra memory required is significant if we have large inputs

Reply:
Thanks for clarifying! Not worth doing it for this PR here then I think; For training people are best off using https://github.com/allenai/OLMo/pull/639